### ISSUE TITLE
protoc: code gen is different between using source files and using descriptor set

### ISSUE THREAD

**Issue by jhump** [protocolbuffers/protobuf#6175]
**What version of protobuf and what language are you using?**
Version: v3.5.1

**What operating system (Linux, Windows, ...) and version?**
OS X

**What did you do?**
I created an extremely simple test file:
```proto
syntax = "proto3";

message Foo {
  uint64 id = 1;
  string name = 2;
  Bar bar = 3 [json_name="barre"];
}

enum Bar {
  VAL0 = 0;
  VAL1 = 1;
  VAL2 = 2;
}
```

I generated this to `Test.java` with the following command:

```bash
protoc test.proto --java_out=.
```

I then produced a descriptor set and used that to generate Java code again:

```bash
protoc test.proto -o test.protoset --include_source_info
protoc test.proto --descriptor_set_in test.protoset --java_out=.
```

**What did you expect to see**

I expected the output from the two code gen approaches to be the same. (And having contributed to this repo and knowing some of the code, I'm a bit surprised, since I know the compiler first parses into a descriptor and then does code gen based on the descriptor -- it's weird that the descriptor just parsed would differ from the descriptor de-serialized from file.)

**What did you see instead?**

Instead, it produced different Java output. When using a protoset, it _always_ emits comments that look as if the `json_name` option were always specified. But when using source, it only includes that comment when the `json_name` option _is actually_ specified.

This smells related to #5587.

**Comment by jhump** [comment]
cc @pedgeio

**Comment by pedgeio** [comment]
I believe the issue is that `json_name` is always populated when it is returned from `--descriptor_set_out`, but when `protoc` calls one of it's builtin plugins internally, it doesn't populate `json_name` unless it is explicitly set (this comment hints at this: https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.cc#L6024).

While this seems minor, this is important - I should be able to do a roundtrip with `protoc`, ie generate directly using it, or generate from a pre-computed `FileDescriptorSet`, and have the same result.

Possible fixes:

- Always populate `json_name` for `FileDescriptorSets` passed to builtin plugins - this immediately fixes everything.
- Remove `json_name` when passing `FileDescriptorSets` given with `--descriptor_set_in` before passing them to builtin plugins if the value of `json_name` is the same as the default computed value for `json_name` for that field - this also would fix the builtin plugin issue.

As an example example, here's the diff between what is generated by an inputted `FileDescriptorSet` to `protoc`, and `protoc` directly, for `AdImageAsset.java` from https://github.com/googleapis/googleapis/blob/master/google/ads/googleads/v1/common/ad_asset.proto:

```diff
01c101
<    * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>    * <code>.google.protobuf.StringValue asset = 1;</code>
111c111
<    * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>    * <code>.google.protobuf.StringValue asset = 1;</code>
121c121
<    * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>    * <code>.google.protobuf.StringValue asset = 1;</code>
454c454
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
464c464
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
478c478
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
498c498
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
516c516
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
538c538
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
556c556
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
568c568
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
583c583
<      * <code>.google.protobuf.StringValue asset = 1[json_name = "asset"];</code>
---
>      * <code>.google.protobuf.StringValue asset = 1;</code>
```

**Comment by pedgeio** [comment]
cc @haon4 (let me know if you don't want cc's for maintainers on issues by the way, if so this will be the only time)

**Comment by pedgeio** [comment]
Just for clarification: note in the above example that @jhump provides, regardless of if there is any `json_name` explicitly set on a field or not, this will happen. Ie this https://github.com/googleapis/googleapis/blob/master/google/ads/googleads/v1/common/ad_asset.proto will produce a diff.

**Comment by pedgeio** [comment]
Proposed another fix in #5587 that I personally think is the best of these options

**Comment by BSBandme** [comment]
I think this issue's root cause is the same as #5587. 
Close this issue here as this is filed later.
